File: C++/0125_validpalindrome.cc
// 2020-08-04 01:09:02

#include <string>

using std::string;

class Solution
{
public:
    bool isPalindrome(string s)
    {
        if (s.length() <= 1)
            return true;
        int i = 0, j = s.length() - 1;
        while (i < j)
        {
            while (i < j && !isalnum(s[i]))
                i++;
            while (i < j && !isalnum(s[j]))
                j--;
            if (i != j && (s[i] & 223) != (s[j] & 223))
                return false;
            i += 1, j -= 1;
        }
        return true;
    }
};

File: Python/0125_valid_palindrome.py
# 2020-08-04 00:02:55

class Solution:
    def isPalindrome(self, s: str) -> bool:
        if s is None or len(s) <= 1:
            return True
        i = 0; j = len(s) - 1
        while i < j:
            while i < j and not s[i].isalnum():
                i += 1
            while i < j and not s[j].isalnum():
                j -= 1
            if (ord(s[i]) & 0b11011111) != (ord(s[j]) & 0b11011111):
                return False
            i += 1; j -= 1
        return True
